swarm/src/behaviour.rs

```diff
@@ -228,15 +242,24 @@ pub trait NetworkBehaviourEventProcess<TEvent> {
 /// in whose context it is executing.
 ///
 /// [`Swarm`]: super::Swarm
+//
+// Note: `TInEvent` is needed to be able to implement
+// [`NetworkBehaviourAction::map_in`], mapping the handler `InEvent` leaving the
+// handler itself untouched.
 #[derive(Debug)]
-pub enum NetworkBehaviourAction<TInEvent, TOutEvent> {
+pub enum NetworkBehaviourAction<
+    TOutEvent,
+    THandler: IntoProtocolsHandler,
+    TInEvent = THandlerInEvent<THandler>,
+> {
```

Comment:
Similar to `THandlerInEvent`, I think we could define `NetworkBehaviourAction` in terms of the `NetworkBehaviour` trait. That would reduce the amount of type parameters down to one.

It would also make it easier to refer to because you can just say `NetworkBehaviourAction<Self>` within the `NetworkBehaviour`.

Thoughts?

Reply:
Ah we would have to give up our `map_XYZ` functions then.

Reply:
Good idea @thomaseizinger which I started to explore. See https://github.com/mxinden/rust-libp2p/pull/5

> Ah we would have to give up our map_XYZ functions then.

Not as far as I can tell with additional trait parameters (each defaulting to a derivative of `TBehaviour`). See PR above.

Unfortunately I don't see a way to _use_ `TBehaviour` in `NetworkBehaviourAction`. Unless I do so, compilation fails.

Let's move the discussion to https://github.com/mxinden/rust-libp2p/pull/5.